src/modules/claim/services/claim-issuance.service.ts

```diff
@@ -0,0 +1,150 @@
+import { Logger } from '../../logger/logger.service';
+import { ClaimIssueDTO, NewClaimIssueDTO } from '../claim.dto';
+import { RoleClaim } from '../entities/roleClaim.entity';
+import { ClaimService } from './claim.service';
+import { ClaimHandleResult } from '../claim-handle-result.dto';
+
+@Injectable()
+export class ClaimIssuanceService {
+  constructor(
+    private readonly roleService: RoleService,
+    private readonly logger: Logger,
+    @InjectRepository(RoleClaim)
+    private readonly roleClaimRepository: Repository<RoleClaim>
+  ) {
+    this.logger.setContext(ClaimIssuanceService.name);
+  }
+
+  /**
+   * Handles claim issuance request saving and updates.
+   * Two scenarios are handled - issue requested claim and issue not-requested claim
+   * @param rq IClaimIssuance request
+   */
+  public async handleClaimIssuanceRequest(
+    rq: IClaimIssuance,
+    previouslyRequestedClaim: RoleClaim | undefined
+  ): Promise<ClaimHandleResult> {
+    if (!previouslyRequestedClaim && (rq.issuedToken || rq.onChainProof)) {
+      return this.handleClaimWithoutPreviousRequest(rq);
+    } else if (
+      previouslyRequestedClaim &&
+      !previouslyRequestedClaim.isRejected &&
+      (rq.issuedToken || rq.onChainProof)
+    ) {
+      return this.handleClaimWithPreviousRequest(rq, previouslyRequestedClaim);
+    }
+
+    return ClaimHandleResult.Failure('claim issuance criteria not met');
+  }
+
+  private async handleClaimWithoutPreviousRequest(
+    rq: IClaimIssuance
+  ): Promise<ClaimHandleResult> {
+    let claimType: string;
+    let claimTypeVersion: string;
+
+    if (rq.issuedToken) {
+      const { claimData } = jwt.decode(rq.issuedToken) as DecodedClaimToken;
+      claimType = claimData.claimType;
+      claimTypeVersion = claimData.claimTypeVersion;
+    } else {
+      claimType = rq.claimType;
+      claimTypeVersion = rq.claimTypeVersion;
+    }
+
+    // TODO: add verification of on-chain proof
+    // if (
+    //   rq.onChainProof &&
+    //   !this.verifyOnChainProof
+    // ) {
+    //   return ClaimHandleResult.Failure('on-chain proof verification failed');
+    // }
+
+    if (!claimType || !claimTypeVersion) {
+      return ClaimHandleResult.Failure('claim type or version not provided');
+    }
+
+    const dto = await NewClaimIssueDTO.create({
+      ...rq,
+      claimType,
+      claimTypeVersion,
+    });
+
+    await this.roleService.verifyEnrolmentIssuer({
+      issuerDID: dto.acceptedBy,
+      claimType: dto.claimType,
+    });
+
+    await this.roleService.verifyEnrolmentPrecondition({
+      claimType: dto.claimType,
+      userDID: dto.requester,
+    });
+
+    await this.createAndIssue(dto, dto.requester);
+
+    return ClaimHandleResult.Success();
+  }
+
+  private async handleClaimWithPreviousRequest(
+    rq: IClaimIssuance,
+    previouslyRequestedClaim: RoleClaim
+  ): Promise<ClaimHandleResult> {
+    await this.roleService.verifyEnrolmentIssuer({
```

Comment:
Exact verification is also in `handleClaimWithoutPreviousRequest`. Let's hoist it

Reply:
It will complicated the code from my perspective. Maybe you can suggest the change here

Reply:
I would do like this:
in `handleClaimIssuanceRequest`
- make sure that `issuedToken` or `onChainProof` is present
- make sure that `claimType` and `claimTypeVersion` are present, otherwise extract them from `issuedToken` or throw
- `verifyEnrollmentIssuer`, `verifyEnrollmentPrecodnditions`
- depending on existence of `previousyRequestedClaim` call `issue` or `createAndIssue`

Reply:
Okay, lets see how it will look like